tools/dm/shard-merge.md

```diff
@@ -0,0 +1,175 @@
+
+分库分表合并同步过程中，可以根据 column values 及下游的表结构构造出相应的 DML 语句，但此时若上游的表执行 DDL 语句进行了表结构变更，则必须对 DDL 语句同步过程进行额外处理，以避免造成由 column values 生成的 DML 语句与下游实际表结构不一致的问题。
+
+以下是一个简化后的例子：
+
+![shard-ddl-example-1](/media/shard-ddl-example-1.png)
+
+在上图的例子中，分表的合库合表过程简化成了上游只有两个 MySQL 实例，每个实例内只有一个表。假设在数据同步开始时，将两个分表的表结构版本记为 schema V1，将 DDL 语句执行完后的表结构版本记为 schema V2。
+
+现在，假设数据同步过程中，DM-worker 内的 sync 单元从两个上游分表收到的 binlog 数据有如下时序：
+
+1. 开始同步时，sync 单元从两个分表收到的都是 schema V1 版本的 DML 语句。
+
+2. 在 t1 时刻，sync 单元收到实例 1 上分表的 DDL 语句。
+
+3. 从 t2 时刻开始，sync 单元从实例 1 收到的是 schema V2 版本的 DML 语句；但从实例 2 收到的仍是 schema V1 版本的 DML 语句。
+
+4. 在 t3 时刻，sync 单元收到实例 2 上分表的 DDL 语句。
+
+5. 从 t4 时刻开始，sync 单元从实例 2 收到的也是 schema V2 版本的 DML 语句。
+
+假设在数据同步过程中，不对分表的 DDL 语句进行额外处理。当实例 1 的 DDL 语句同步到下游后，下游的表结构会变更成为 schema V2 版本。但在 t2 到 t3 这段时间内，sync 单元从实例 2 上收到的仍是 schema V1 版本的 DML 语句。当尝试把这些 schema V1 版本的 DML 语句同步到下游时，就会由于 DML 语句与表结构的不一致而发生错误，从而无法正确同步数据。
+
+### 实现原理
+
+基于上述例子，本部分介绍了 DM 在合库合表过程中进行 DDL 同步的实现原理。
+
+![shard-ddl-flow](/media/shard-ddl-flow.png)
+
+在这个例子中，DM-worker-1 负责同步来自 MySQL 实例 1 的数据，DM-worker-2 负责同步来自 MySQL 实例 2 的数据，DM-master 负责协调多个 DM-worker 间的 DDL 同步。
+
+从 DM-worker-1 收到 DDL 语句开始，简化后的 DDL 同步流程为：
+
+1. 在 t1 时刻，DM-worker-1 收到来自 MySQL 实例 1 的 DDL 语句，自身暂停该 DDL 语句对应任务的 DDL 及 DML 数据同步，并将 DDL 相关信息发送给 DM-master。
+
+2. DM-master 根据收到的 DDL 信息判断得知需要协调该 DDL 语句的同步，于是为该 DDL 语句创建一个锁，并将 DDL 锁信息发回给 DM-worker-1，同时将 DM-worker-1 标记为这个锁的 owner。
+
+3. DM-worker-2 继续进行 DML 语句的同步，直到在 t3 时刻收到来自 MySQL 实例 2 的 DDL 语句，自身暂停该 DDL 语句对应任务的数据同步，并将 DDL 相关信息发送给 DM-master。
+
+4. DM-master 根据收到的 DDL 信息判断得知该 DDL 语句对应的锁信息已经存在，于是直接将对应锁信息发回给 DM-worker-2。
+
+5. 根据任务启动时的配置信息、上游 MySQL 实例分表信息、部署拓扑信息等，DM-master 判断得知自身已经收到了来自待合表的所有上游分表的 DDL 语句，于是请求 DDL 锁的 owner（DM-worker-1）向下游同步执行该 DDL。
+
+6. DM-worker-1 根据第二步收到的 DDL 锁信息验证 DDL 语句执行请求；向下游执行 DDL，并将执行结果反馈给 DM-master；若 DDL 语句执行成功，则自身开始继续同步后续的（从 t2 时刻对应的 binlog 开始的）DML 语句。
+
+7. DM-master 收到来自 owner 执行 DDL 语句成功的响应，于是请求在等待该 DDL 锁的所有其他 DM-worker（DM-worker-2）忽略该 DDL 语句，直接继续同步后续的（从 t4 时刻对应的 binlog 开始的）DML 语句。
+
+根据上面的流程，可以归纳出 DM 协调多个 DM-worker 间 sharding DDL 同步的特点：
+
+- 根据任务配置与 DM 集群部署拓扑信息，DM-master 内部会建立一个逻辑 sharding group 来协调 DDL 同步，group 中的成员即为负责处理该同步任务拆解后的各子任务的 DM-worker。
+
+- 各 DM-worker 从 binlog event 中收到 DDL 语句后，会将 DDL 信息发送给 DM-master。
+
+- DM-master 根据来自 DM-worker 的 DDL 信息及 sharding group 信息创建或更新 DDL 锁。
+
+- 如果 sharding group 的所有成员都收到了某一条相同的 DDL 语句，则表明上游分表在该 DDL 执行前的 DML 语句都已经同步完成，此时可以执行该 DDL 语句，并继续后续的 DML 同步。
+
+- 上游分表的 DDL 在经过 [table router](/tools/dm/data-synchronization-features.md#table-routing)  转换后，需要与在下游执行的 DDL 语句保持一致，因此仅需 DDL 锁的 owner 执行一次该 DDL 语句即可，其他 DM-worker 可直接忽略对应的 DDL 语句。
+
+在上面的示例中，每个 DM-worker 对应的上游 MySQL 实例中只有一个待合并的分表。但在实际场景下，一个 MySQL 实例可能有多个分库内的多个分表需要进行合并，这种情况下，sharding DDL 的协调同步过程将更加复杂。
+
+假设同一个 MySQL 实例中有 table_1 和 table_2 两个分表需要进行合并：
+
+![shard-ddl-example-2](/media/shard-ddl-example-2.png)
+
+在这个例子中，由于数据来自同一个 MySQL 实例，因此所有数据都是从同一个 binlog 流中获得，时序如下：
+
+1. 开始同步时，DM-worker 内的 sync 单元从两个分表收到的数据都是 schema V1 版本的 DML 语句。
```

Comment:
syncer？ or binlog replication

Reply:
@csuzhangxc told me that I should use "binlog replication unit (sync)" and "Syncer" is deprecated. So how about using "binlog 同步单元" rather than "sync 单元"?